sources/org.osbuild.librepo

```diff
@@ -0,0 +1,281 @@
+        # this is currently using the CURL names, librepo is using
+        # curl under the hood so this is okay-ish, but maybe we should
+        # generalize the name
+        key = os.getenv("OSBUILD_SOURCES_CURL_SSL_CLIENT_KEY")
+        cert = os.getenv("OSBUILD_SOURCES_CURL_SSL_CLIENT_CERT")
+        if not (key and cert):
+            raise RuntimeError(f"mtls secrets required but key ({key}) or cert ({cert}) not defined")
+        handle.sslcacert = os.getenv("OSBUILD_SOURCES_CURL_SSL_CA_CERT")
+        handle.sslclientcert = cert
+        handle.sslclientkey = key
+
+    # This gets called when done
+    # data comes from cbdata
+    # status is librepo.TRANSFER_*
+    #   librepo.TRANSFER_SUCCESSFUL
+    #   librepo.TRANSFER_ALREADYEXISTS
+    #   librepo.TRANSFER_ERROR
+    def _endcb(self, data, status, msg):
+        """Callback for librepo transfers
+
+        data is the name/path of the package
+        status is a librepo TRANSFER_* status code
+        msg is a status message or error
+
+        TRANSFER_ERROR is returned if all mirrors are tried and it cannot download
+        the file.
+        """
+        if status == librepo.TRANSFER_ERROR:
+            self.errors.append(f"{data}: {msg}")
+        # Workaround the lack of structured progress reporting from
+        # stages/sources. See curl source for more details, once we
+        # have something better this block can be removed.
+        if status == librepo.TRANSFER_ERROR:
+            print(f"Error downloading {data}")
+        elif status == librepo.TRANSFER_SUCCESSFUL:
+            print(f"Downloaded {data}")
+        elif status == librepo.TRANSFER_ALREADYEXISTS:
+            print(f"Already downloaded {data}")
+        else:
+            print(f"Unknown status {status} for {data}")
+
+    def make_pkg_target(self, handle, dest, path, checksum):
+        """Return a librepo.PackageTarget populated with the package data
+
+        This specifies what to download, where to save it, the checksum, etc.
+        """
+        chksum_type, checksum = checksum.split(":")
+        return librepo.PackageTarget(
+            path,
+            handle=handle,
+            checksum_type=self.CHKSUM_TYPE[chksum_type],
+            checksum=checksum,
+            dest=dest,
+            cbdata=path,
+            endcb=self._endcb)
+
+    # pylint: disable=too-many-branches
+    def fetch_all(self, items: Dict) -> None:
+        """Use librepo to download the packages"""
+        # Organize the packages by the mirror id
+        packages_from_mirror = {}
+        for item_id, pkg in items.items():
+            if pkg["mirror"] not in self.options["mirrors"]:
+                raise RuntimeError(f'Missing mirror: {pkg["mirror"]}')
+
+            if pkg["mirror"] not in packages_from_mirror:
+                packages_from_mirror[pkg["mirror"]] = [(pkg["path"], item_id)]
+            else:
+                packages_from_mirror[pkg["mirror"]].append((pkg["path"], item_id))
```

Comment:
Nitpick: I tend to like the `setdefault()` method of `dict`, which would allow you to rewrite this as:

```python
packages_from_mirror.setdefault(pkg["mirror"], []).append((pkg["path"], item_id))
```

Reply:
Thank you! Love it!

Reply:
I applied your suggestions, sadly it means I need the review again :)